main/src/ca/uwaterloo/flix/language/phase/unification/BoolMinimization.scala

```diff
@@ -0,0 +1,482 @@
+/*
+ *  Copyright 2022 Jonathan Lindegaard Starup
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.language.ast.{Kind, Rigidity, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+import scala.annotation.tailrec
+
+object BoolMinimization {
+
+  sealed trait Formula
+
+  object Formula {
+
+    case object True extends Formula
+
+    case object False extends Formula
+
+    case class Var(v: Type.KindedVar) extends Formula
+
+    case class Not(term: Formula) extends Formula
+
+    case class And(terms: List[Formula]) extends Formula
```

Comment:
I think we need to do one of two things:

- Either allow `And(Nil)` and consider it equivalent to `True` whenever we see it, or
- Change the constructor to `And(hd: Formula, tl: List[Formula]` to ensure its non-empty.